### PR TITLE
*: refine context usage to avoid potential blocks

### LINKED ISSUE #1279: query-status does return some abnormal tasks
## Bug Report

Please answer these questions before submitting your issue. Thanks!

1. What did you do? If possible, provide a recipe for reproducing the error.

    `query-status`

2. What did you expect to see?

    all normal or abnormal tasks

3. What did you see instead?

    some abnormal tasks are missing in return

image

image

[dm-worker-goroutine.txt](https://github.com/pingcap/dm/files/5508496/dm-worker-goroutine.txt)

`Syncer.Status` may be blocked for `SHOW MASTER STATUS` for a long time, but gRPC from DM-master to DM-worker will become timeout in 30s.

4. Versions of the cluster

    - DM version (run `dmctl -V` or `dm-worker -V` or `dm-master -V`):

        ```
        v2.0.0
        ```

### PR DESCRIPTION
**Pull request by csuzhangxc**
### What problem does this PR solve? 

refine context usage to avoid potential blocks, like:

fix #1279 

### What is changed and how it works?

add context for DB operations.

**some suggestions for later dev:**

- **always try to use `QueryContext`/`ExecContext` rather than `Query`/`Exec` for DB operations**
- **try to avoid using `context.Background` for most cases except in testing code**
- combine `context.Context` and `logger` into a `tcontext.Context` may be not a good idea, becase `context.Context` often takes effect in a single process workflow but `logger` often takes effect in a whole component
- some `tcontext.Context` may still need to be refined later

### Check List 

Tests 

 - Unit test
 - Integration test

Code changes

 - Has exported function/method change
 - Has interface methods change

Side effects

 - Increased code complexity

Related changes

 - Need to cherry-pick to the release branch
 - Need to be included in the release note

### CODE REVIEW

File: tests/all_mode/run.sh [hunk 1]
```diff
@@ -52,11 +52,60 @@ function test_session_config(){
...
+    sed -i "/relay-binlog-name/i\relay-dir: $WORK_DIR/worker2/relay_log" $WORK_DIR/source2.yaml
+    dmctl_operate_source create $WORK_DIR/source1.yaml $SOURCE_ID1
+    dmctl_operate_source create $WORK_DIR/source2.yaml $SOURCE_ID2
+
+    # start DM task only
+    cp $cur/conf/dm-task.yaml $WORK_DIR/dm-task.yaml
+    sed -i "s/name: test/name: $ILLEGAL_CHAR_NAME/g" $WORK_DIR/dm-task.yaml
+    dmctl_start_task "$WORK_DIR/dm-task.yaml" "--remove-meta"
+    
+    # `query-status` timeout
```
Comment (lance6716) [inline tests/all_mode/run.sh:86]: how about testing it doesn't cost much time

**Review by lance6716 — COMMENTED** [review]
rest LGTM

File: tests/all_mode/run.sh (hunk 1 above)
Comment (csuzhangxc) [reply tests/all_mode/run.sh:86]: added testing in https://github.com/pingcap/dm/pull/1285/commits/5f72910f8f8cb5e940b874f7ae9ae44cb6046576.

**Comment by lance6716** [comment]
LGTM

**Review by GMHDBJD — APPROVED** [review]
LGTM

**Pull request merged**